src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+            {
+                // Here's the entire process for changing namespace:
+                // 1. Change the namespace declaration, fix references and add imports that might be necessary.
+                // 2. Explicitly merge the diff to get a new solution.
+                // 3. Remove added imports that are unnecessary.
+                // 4. Do another explicit diff merge based on last merged solution.
+                //
+                // The reason for doing explicit diff merge twice is so merging after remove unnecessaty imports can be correctly handled.
+
+                var solutionAfterNamespaceChange = _state.Solution;
+                var referenceDocuments = PooledHashSet<DocumentId>.GetInstance();
+
+                foreach (var id in _state.DocumentIds)
+                {
+                    var result = await ChangeNamespaceToMatchFoldersAsync(solutionAfterNamespaceChange, id, cancellationToken).ConfigureAwait(false);
+                    solutionAfterNamespaceChange = result.Item1;
+                    referenceDocuments.AddRange(result.Item2);
+                }
+
+                var solutionAfterFirstMerge = await MergeDiffAsync(_state.Solution, solutionAfterNamespaceChange, cancellationToken).ConfigureAwait(false);
+
+                // After changing documents, we still need to remove unnecessary imports related to our change.
+                // We don't try to remove all imports that might become unnecessary/invalid after the namespace change, 
+                // just ones that fully matche the old or new namespace.
+                // For example, if we are changing namespace `Foo.Bar` to `A.B`, the using of name `Bar` below would remain 
+                // untouched:
+                //
+                //      namespace Foo
+                //      {
+                //          using Bar;
+                //          ~~~~~~~~~
+                //      }
+                //
+                // Also, because we may have added different imports to document that triggered the refactoring
+                // and the documents that reference affected types declared in changed namespace, we try to remove
+                // unnecessary imports separately.
+                var solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterFirstMerge,
+                    _state.DocumentIds, 
+                    CreateAllContainingNamespaces(OldNamespaceParts), 
+                    cancellationToken).ConfigureAwait(false);
+
+                solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterImportsRemoved, 
+                    referenceDocuments.ToImmutableArray(), 
+                    ImmutableArray.Create(_state.DeclaredNamespace, _state.TargetNamespace), 
+                    cancellationToken).ConfigureAwait(false);
```

Comment:
sorry, it's unclear to me why we need to do RemoveUnnecessaryImports in two passes.  Could you clarify?  Or point to the comment i've missed that explains why?

Reply:
There's no comment, I thought the intention would be clear with just the code.

The first pass remove usings from document with namespace declaration change, and the second pass remove usings from all documents referencing types from changed namespace. The usings to be removed are different between declaring document and reference documents, and I want to minimize the impact on the code. (e.g. I don't want to remove an unnecessary using if it was there before the refactoring)

Reply:
That being said, I don't have a strong opinion on this, since we are not removing all unnecessary usings, just maybe a few of the containing namespaces of change namespace. So if you think it's OK and make things faster. I can combine them into one single pass.

Reply:
thanks for the explanation.  I'm fine with this.  It was just a little unclear when i first read it :)